### ISSUE TITLE
Import-Module IISAdministration Fails With "Could not load file or assembly 'Microsoft.Web.Administration..."

### ISSUE THREAD

**Issue by tparikka** [PowerShell/PowerShell#17012]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

On Windows, run the following script in PowerShell 7.2.1:

```powershell
Install-Module -Name IISAdministration
Import-Module IISAdministration
Get-Module -ListAvailable
```

I have confirmed on a VM running 19042 that the Windows 10 build is not a factor as far as I can tell, because on a 20H2 VM I still see the problem behavior. I have also enabled IIS Management Console and IIS Management Scripts and Tools. The bulk of the documentation I could find referred to earlier versions of PowerShell 7 not supporting IISAdministration, but this doesn't seem to apply to 7.2.1 - if it did globally, I wouldn't have myself and multiple other team members running the scripts with no issues. I appreciate any help!

I did see in issue #13818 the reporter was told to look in the https://github.com/PowerShell/PowerShellModuleCoverage repo and I reported the issue there as well, but then I saw the repo appears to be defunct. Its last commit was literally 4 years ago. ALl the other issues I can find reported for IISAdministration and PowerShell Core / 7 are at least two years old and refer specifically to earlier versions of PowerShell being incompatible, but I can see from the fact that the module works on several developer machines in my department that is no longer true.

### Expected behavior

The import should succeed and the module should be listed:

```console
Script     1.1.0.0               IISAdministration                   Core,Desk {Get-IISServerManager, Set-IISConfigAtt…
```

### Actual behavior

The import fails with the following error:

```console
Import-Module: Could not load file or assembly 'Microsoft.Web.Administration, Version=7.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.

The system cannot find the file specified.
```

The Get-Module list shows no cmdlets associated with IISAdministration:

```none
Script     1.1.0.0               IISAdministration                   Core,Desk
```

### Environment data

On the machine exhibiting the problem behavior:

```powershell
PS C:\Users\thoma> $PSVersiontable

Name                           Value
----                           -----
PSVersion                      7.2.1
PSEdition                      Core
GitCommitId                    7.2.1
OS                             Microsoft Windows 10.0.19044
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

On the machine working successfully:

```powershell
PS C:\Users\thomas.parikkaADM> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.2.1
PSEdition                      Core
GitCommitId                    7.2.1
OS                             Microsoft Windows 10.0.19042
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by johnmbaughman** [comment]
Another year later and still no response...

I found an issue with at least one installation of IIS and this module. For some reason, in the `C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules` folder the IISAdministration module was installed possibly by the IIS feature installation, and the files are set to read-only and execute for permissions outside of the TrustedInstaller owner permissions. If it is installed with `Install-Module`, the folder `C:\Program Files\PowerShell\Modules` is used and installed into the versioned folder there.

The workaround I figured out is a multi-step process. 

1. Install the module: `Install-Module -Name IISAdministration -Scope AllUsers -AllowClobber`
2. Copy the version folder (current version as of this comment: `1.1.0.0`) to `C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules\IISAdministration`
3. Uninstall the module: `Uninstall-Module -Name IISAdministration -AllVersions` 
4. Import the module with the `-RequiredVersion` parameter: `Import-Module -Name IISAdministration -RequiredVersion 1.1.0.0`

This works perfectly, although very annoying. Note that if you do not use the `-RequiredVersion` parameter, you will get the previous version. This version and the previous version do not show on `Get-Module -ListAvailable` because, for some reason, PS7+ ignores the PSModulePath value of `C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules` (bug or expected behavior? I'm leaning towards the latter as you might get some very unexpected results from very old modules in PS7+.)

**Comment by johnmbaughman** [comment]
As a side note, when importing the module you get a message to use the `-SkipEditionCheck` parameter. This results in the aforementioned `Could not load assembly` error.

![image](https://user-images.githubusercontent.com/1634414/223563425-14d258c7-e0dd-41d9-bd72-675f1d216ed4.png)

Using this parameter fails to load the module as expected. Ignoring the initial warning message appears to still load the module and allow cmdlets to work.

**Comment by tparikka** [comment]
I was issued a new laptop by work last week, Windows 10 22H2, and PS 7.3.3:

```ps
PowerShell 7.3.3
PS C:\Users\thomas.parikka> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.3.3
PSEdition                      Core
GitCommitId                    7.3.3
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

Same issue occurs, was forced again to delete the IisAdministration module from C:\Program Files\PowerShell\Modules before anything would function.

**Comment by doctordns** [comment]
At the end of the day, this is an issue for the IIS Team. They need to update the module to work properly with .NET Core. Until then, the best you will probably get is using the module via the Windows PowerShell compatibility mechanism. The fix here is probably external (and frankly, given the Windows Division;'s apparent lack of enthusiasm) will probably never happen.  I'd love to be wrong, but the IIS and other teams have known about PowerShell for half a decade and have done little to make the various modules directly useable in PowerShell 7. 

In my latest PowerShell book, I gave up on the notion that MS would change things like this, so the last chapter is how to use WSUS with PowerShell 7. Not pretty, but here are the scripts: https://github.com/doctordns/PacktPS72/tree/main/Scripts/Ch%2014%20-%20WSUS

**Comment by tparikka** [comment]
@doctordns Does the IIS team have a GitHub where an issue like this would better belong? It seems wild to me that this would be left by the wayside given Microsoft as a whole is moving past .NET Framework in every other respect.

**Comment by doctordns** [comment]
I have retested this today and here is what I';ve found. If I load the Web-Server in Windows Server 2022 and  include the management tools: as expected, the IISAdministration module is present, can be loaded, and works as designed. The IISAdministration module on ther PS Gallery appears to be the very same module - same Guid in the manifest, for example. I downloaded this and can install the module and use it.

tl:dr - I can not currently repo the issue 

I also note an issue at your end: namely IISAdministration module is being loaded using the compatibility mechanism. On all my servers, using PowerShell 7.2.6, the module loads natively.

**Comment by TobiasPSP** [comment]
The cmdlet working group couldn't repro the issue on a Windows 11 test system. If issues exist, they are related to this particular module and it requirements that go beyond PowerShells'  built-in backwards compatibility machanism. The module was last updated in 2017. To solve the issue, you'd have to approach the module author.

**Comment by george-andrei** [comment]
> Another year later and still no response...
> 
> I found an issue with at least one installation of IIS and this module. For some reason, in the `C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules` folder the IISAdministration module was installed possibly by the IIS feature installation, and the files are set to read-only and execute for permissions outside of the TrustedInstaller owner permissions. If it is installed with `Install-Module`, the folder `C:\Program Files\PowerShell\Modules` is used and installed into the versioned folder there.
> 
> The workaround I figured out is a multi-step process.
> 
> 1. Install the module: `Install-Module -Name IISAdministration -Scope AllUsers -AllowClobber`
> 2. Copy the version folder (current version as of this comment: `1.1.0.0`) to `C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules\IISAdministration`
> 3. Uninstall the module: `Uninstall-Module -Name IISAdministration -AllVersions`
> 4. Import the module with the `-RequiredVersion` parameter: `Import-Module -Name IISAdministration -RequiredVersion 1.1.0.0`
> 
> This works perfectly, although very annoying. Note that if you do not use the `-RequiredVersion` parameter, you will get the previous version. This version and the previous version do not show on `Get-Module -ListAvailable` because, for some reason, PS7+ ignores the PSModulePath value of `C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules` (bug or expected behavior? I'm leaning towards the latter as you might get some very unexpected results from very old modules in PS7+). `Import-Module` does not ignore that path.

@doctordns Unfortunately, I did not had any luck with this method. I cannot replace the existing files in `C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules\IISAdministration`

All my tests are done in _Windows Server 2019 v1809 build 17763.4499_

This version of the OS come with version 1.0 of `IISAdministration` and for multiple reasons I need the version `1.1.0.0` 
The command `Import-Module IISAdministration -RequiredVersion 1.1.0.0` (With or without `-SkipEditionCheck`) fail. So I need to install the latest version present in the PowerShell Gallery. 

As mentioned above, install the module in PS7 with the following command: 

```powershell
Install-Module -Name IISAdministration -Scope AllUsers -AllowClobber
```
After the installation:

```powershell
Import-Module IISAdministration -RequiredVersion 1.1.0.0
Import-Module: Could not load file or assembly 'Microsoft.Web.Administration, Version=7.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.
```

To solve this, copy `C:\Windows\System32\inetsrv\Microsoft.Web.Administration.dll` in the directory where IISAdministration 1.1.0.0 is installed: `C:\Program Files\PowerShell\Modules\IISAdministration\1.1.0.0`

Now import the module again and it will work: 

```powershell
PS C:\Windows\System32> Import-Module IISAdministration -RequiredVersion 1.1.0.0
PS C:\Windows\System32> Get-Module -List IISAdministration

    Directory: C:\Program Files\PowerShell\Modules

ModuleType Version    PreRelease Name                                PSEdition ExportedCommands
---------- -------    ---------- ----                                --------- ----------------
Script     1.1.0.0               IISAdministration                   Desk
```

I hope this can help.

**Comment by sir-wilhelm** [comment]
I was also getting `Assembly with same name is already loaded` errors, but trying to get commands from the module also failed:
```powershell
Import-Module: C:\Users\<redacted>.ps1:47
Line |
  47 |      Import-Module -Name IISAdministration
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Assembly with same name is already loaded
     
Get-Command: C:\Users\<redacted>.ps1:48
Line |
  48 |      Get-Command -Module IISAdministration -Name Get-IISAppPool
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | The term 'Get-IISAppPool' is not recognized as a name of a cmdlet, function, script file, or executable program.
     | Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
WARNING: Failed to stop the Revenue Cycle Dashboard Pool The 'Get-IISAppPool' command was found in the module 'IISAdministration', but the module could not be loaded due to the following error: [Assembly with same name is already loaded]
For more information, run 'Import-Module IISAdministration'.
```

Flipping the import to use the `-UseWindowsPowerShell` flag resolved the issue for me:
```
Import-Module -Name IISAdministration -UseWindowsPowerShell
```
